Issue #620 PR2 break dependency on httpclient and replace with okhttp3

xsteelej opened the pull request:
Closes #620 

<!-- 
Thank you for contributing to ODK Collect!

Before sending this PR, please read
https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md
-->

#### What has been done to verify that this works as intended?
I've tested this on a local VM Aggregate instance with both Digest and Basic authentication.
However, this has not yet been tested with an Ona server and therefore I'm not sure if this implementation suffers the same issue as described in this issue https://github.com/opendatakit/collect/issues/2592.

#### Why is this the best possible solution? Were any other approaches considered?
The use of okhttp3 seemed to be the most obvious choice due to its popularity and ease of use. Most of the ground work has been done with https://github.com/opendatakit/collect/pull/2257

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
This change should not have any affect on users.

#### Do we need any specific form for testing your changes? If so, please attach one.
No
#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

xsteelej committed:
Update README.md

xsteelej committed:
Merge remote-tracking branch 'upstream/master'

# Conflicts:
#	README.md

xsteelej committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Added okhttp3 to the project and created a new OkHttpConnection which implements OpenRosaHttpInterface.

jrsteele09 committed:
Re-added mockwebserver to re-enable the AndroidTest suite.

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge branch 'master' into issue_#620_PR2_break_dependency_on_httpclient

* master:
  Continue image load on orientation change (#2482)

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge branch 'master' into issue_#620_PR2_break_dependency_on_httpclient

* master:
  Expose instance uploader activity to external apps (#2461)
  Handle READ_PHONE_STATE permission, set targetSdk to 23 (#2506)
  Display warning dialog for current() predicate
  Fix crash when Google account is no longer available (#2483)

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge branch 'master' into issue_#620_PR2_break_dependency_on_httpclient

* master:
  Revert Google Cloud SDK to v216 to fix gcloud crash. (#2549)
  Improved the comment
  Retain Firebase perf files (#2548)
  Credentials authentication fix (#2539)
  Add or improve analytics for audit logging, form updates, servers, external apps and external CSV features (#2545)
  Fixed a bug when scrolling to a widget
  Enable the vertical scrollbar in recyclerView
  80% of screen size for RecyclerView may leave some empty space so 90% is a better choice
  Reimplemented selectMulti widgets using RecyclerView
  Separate out search and search() appearances (#2540)
  Log Google Analytics event for protocol on server URL change (#2519)
  Added Google Analytics event to Send Finalized Form (#2544)
  Update September beta translations
  Implemented RecyclerView to improve selectOne speed (#2513)
  Replace download service with activity (#2463)
  Simplify gradle commands to increase efficiency (#2530)

jrsteele09 committed:
Added okhttp-digest to the Project and modified OkHttpConnection to add credentials to the OkHttpClient builder.

jrsteele09 committed:
Fixed androidTests that were broken due to the change over to http3

jrsteele09 committed:
Removed HttpClientConnection and modified the tests to work with OkHttpConnection

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge branch 'master' into issue_#620_PR2_break_dependency_on_httpclient

* master:
  Strip out the search() appearance/function (#2591)
  Save HttpContext between requests (#2594)
  Fixed: attempt to re-open an already-closed object (#2584)
  Fixed Fatal Exception: java.lang.StringIndexOutOfBoundsException: length=0; index=0 (#2582)
  Fixed java.lang.RuntimeException: Can't create handler inside thread that has not called Looper.prepare() (#2580)
  Fixed android.view.WindowManager: Unable to add window -- token android.os.BinderProxy@a68b01e is not valid (#2578)
  Fix attempt to verify if null dialog is showing (#2587)
  Remove Firebase performance monitoring (#2546)
  Update September translations
  Support rank questions in field-list (#2556)
  Support readOnly option in SelectMap widgets (#2559)
  Support read only for rank (#2555)
  Suuport readOnly option in the ArbitraryWidget (#2561)
  App crashes on Selfie video widget if microphone permissions are not set (#2566)

# Conflicts:
#	collect_app/src/main/java/org/odk/collect/android/http/HttpClientConnection.java

jrsteele09 committed:
Fixed a PMD error

jd-alexander commented:
@xsteelej currently checking this out! I am seeing alot of `Merge remote-tracking branch 'upstream/master'` updates from master. Could all of these be squashed?
Probably doesn't matter though, since this whole PR might be squashed anyways :smile:

lognaturel commented:
Lovely to see so much code going away! You will also have the honor of removing `implementation fileTree(dir: 'libs', include: '*.jar')` from the gradle file and the httpclient jar from that libs directory. 🎉 Please include that in this PR!

> therefore I'm not sure if this implementation suffers the same issue as described in this issue

I haven't looked in detail but probably not. It was related to a quirk in how the previous implementation saved header data.

I will aim to get this reviewed in the next couple of days unless someone else beats me to it so that it can go through QA and be merged early in our release cycle. That way folks with other servers like Ona can have lots of time to stress-test it.

> this whole PR might be squashed anyways

Agreed this history looks like a good candidate for squashing.

jd-alexander commented on collect_app/build.gradle:
```diff
@@ -298,7 +301,6 @@ dependencies {
     }
 
     androidTestImplementation "com.squareup.okhttp3:mockwebserver:3.9.0"
```
the `mockwebserver` dependency should have the same version as the main `okhttp` dependency. Would be good if they simply shared the version via a variable.

xsteelej replied:
Thanks, updated the version number to use a variable

jd-alexander commented on collect_app/src/androidTest/java/org/odk/collect/android/utilities/CollectServerClientTest.java:
```diff
@@ -21,19 +21,19 @@
-            assertTrue(r.getHeader("User-Agent").matches("Dalvik/.* org.odk.collect.android/.*"));
-        }
+        // then
+        RecordedRequest r = nextRequest();
```
`r` can simply be called `recordedRequest` 😄

xsteelej replied:
Agreed - I love Android studio's rename refactor 😃

jd-alexander commented on collect_app/src/androidTest/java/org/odk/collect/android/utilities/CollectServerClientTest.java:
```diff
@@ -21,19 +21,19 @@
-            // when
-            doRequest(url("/some-path"));
+    @Test
+    public void httpRequests_shouldHaveUseragentHeader() throws Exception {
```
Should `agent` be capitalized here for more clarity? What do you think?

xsteelej replied:
👍

lognaturel commented on collect_app/src/androidTest/java/org/odk/collect/android/tasks/InstanceServerUploaderTest.java:
```diff
@@ -68,13 +69,13 @@ public void shouldUploadAnInstance() throws Exception {
             assertMatches("Dalvik/.* org.odk.collect.android/.*", r.getHeader("User-Agent"));
             assertEquals("1.0", r.getHeader("X-OpenRosa-Version"));
-            assertEquals("gzip,deflate", r.getHeader("Accept-Encoding"));
+            assertEquals("gzip", r.getHeader("Accept-Encoding"));
```
This means previously zlib-encoded responses were accepted but now they aren't anymore, right? Could this have impact on certain server implementations?

xsteelej replied:
After a bit of googling, I came across this [okhttp 3: how to decompress gzip/deflate response manually using Java/Android](https://stackoverflow.com/questions/51901333/okhttp-3-how-to-decompress-gzip-deflate-response-manually-using-java-android/51902833)

It looks like to support "deflate" then I'll have to add an interceptor.

lognaturel commented on collect_app/src/androidTest/java/org/odk/collect/android/tasks/InstanceServerUploaderTest.java:
```diff
@@ -68,13 +69,13 @@ public void shouldUploadAnInstance() throws Exception {
-                            "Content-Type: text/xml; charset=ISO-8859-1\r",
-                            "Content-Transfer-Encoding: binary\r",
+                            "Content-Type: text/xml\r",
+                            "Content-Length: 20\r",
```
What are the implications of these header changes?

xsteelej replied:
That's a good question, I'm not sure what effect removing "Content-Transfer-Encoding: binary" would be - It seems to be a one of the differences between the two libraries - I.e. httpclient added it automatically.

xsteelej replied:
Is there any way @lognaturel that we can check server side as to what effect this has?

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/http/OkHttpConnection.java:
```diff
@@ -0,0 +1,352 @@
+        }
+    }
+
+    private Request getRequest(@NonNull URI uri) throws MalformedURLException {
```
I generally expect method names to start with verbs so this particular one doesn't do what I expect. I'd rename this one and the following two to `buildGetRequest`, `buildHeadRequest`, `buildPostRequest`.

xsteelej replied:
Agreed, updated

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/http/OkHttpConnection.java:
```diff
@@ -0,0 +1,352 @@
+        Map<String, String> responseHeaders = new HashMap<>();
+        Headers headers = response.headers();
+
+        if (headers.size() > 0) {
```
This check isn't needed, right? If the size is 0, the first loop bound check will fail and nothing will happen, I think.

xsteelej replied:
Of course yes, didn't spot that 😄

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/http/OkHttpConnection.java:
```diff
@@ -0,0 +1,352 @@
+
+    @NonNull
+    @Override
+    public HttpHeadResult head(@NonNull URI uri, @Nullable HttpCredentialsInterface credentials) throws Exception {
```
Consider something more verb-y like `executeHeadRequest` (and `executeGetRequest` above)? That's been bothering me since the prior refactor but I didn't have a great suggestion.

xsteelej replied:
I've renamed them to executeGetRequest and executeHeadRequest respectively.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/http/OkHttpConnection.java:
```diff
@@ -0,0 +1,352 @@
+
+            discardEntityBytes(response);
+
+        } catch (IOException | IllegalStateException e) {
```
As far as I can tell, these catch blocks don't add anything. It seems likely that the raw messages for `MalformedURLException`s and `IllegalStateException`s are more useful than the ones that get thrown here. And I don't think pre-pending "Generic Exception" to an exception string really helps with troubleshooting. I guess the possible exception would be an exception that doesn't have a message? That doesn't seem very likely to me.

Seems like a good opportunity to clean them up and remove even a little bit more code.

lognaturel replied:
If you agree removing them is ok, then I'd also reorder the statements in this method so they match the structure of the get request one. That is, start by building the client, then the request, etc. That way it's easier to see that they're really quite analogous.

xsteelej replied:
Agreed, removed and reordered.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/http/OkHttpConnection.java:
```diff
@@ -0,0 +1,352 @@
+                    return messageParser;
+                }
+            } catch (IOException e) {
+                Timber.e(e);
```
Same comment as above about the catch block. This one does some logging but I think the exception will always have something to do specifically with user data so it's not very helpful to send to crash reporting. Plus, I think these exceptions eventually get logged by one of the callers.

Once this gets thrown out, similar to suggestions above, I'd add an `executePostRequest` method to the interface that follows the same pattern as the get and head methods. The method before it could be `buildMultipartBody` or something like that. I haven't thought through whether it makes sense to keep `uploadSubmissionFile` as part of the interface or split it into the two smaller methods.

xsteelej replied:
Agreed about the Catch block, I also refactored and created an executePostRequest  method as you suggested. This also lead me down the path of separating the functionality in ResponseMessageParser; I created a new class: HttpPostResult to contain the results of the post, leaving ResponseMessageParser to parse the XML - I think it's cleaner that way.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/http/OkHttpConnection.java:
```diff
@@ -0,0 +1,352 @@
+        boolean first = true;
+        int fileIndex = 0;
+        int lastFileIndex;
+        while (fileIndex < fileList.size() || first) {
```
This whole section for building the parts could use some refactoring. Not urgent and a good start might be to add a comment or two. Things like, on the `first` flag, "Always add the submission XML, even if there are no attachments" and on the `while` loop, "split the files to send into parts below the part length threshold"

xsteelej replied:
Agreed, I refactored it a little bit - but was more concerned with keeping the logic as close as possible. Could this be done under a new Issue number / PR?

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/http/OkHttpConnection.java:
```diff
@@ -0,0 +1,352 @@
+            }
+        }
+
+        return messageParser;
```
Wondering whether `ResponseMessageParser` should really be renamed to `HttpPostResult` by analogy to the other two methods above. And I also believe any HTTP response could provide a custom `message` though of course it's much more useful in response to a post: https://docs.opendatakit.org/openrosa-http/#http-responses. That could be filed as a separate issue.

xsteelej replied:
See comment above, I refactored ResponseMessageParser and created a HttpPostResult.

lognaturel commented:
This closes #2520 as well, right?

The one thing I'd really like to see addressed in the PR description is that there is a change in headers provided. I think that's ok but it seems like a source of risk. In particular, I wonder about the removal of `deflate` and whether that could have consequences for certain server implementations. Also, I see that the original code used preemptive basic auth. Is this handled somewhere? Does it matter?

@xsteelej did you see in the [docs](https://square.github.io/okhttp/3.x/okhttp/okhttp3/OkHttpClient.html) that "OkHttpClients should be shared"? That would be a good improvement to have in here. This is especially meaningful if someone is sending multiple submissions at the same time (possibly 100s at the end of the day, for example). The same client should be reused. 

From some quick reading, I think using `Interceptor`s can make sharing e.g. headers between requests easier and that could probably replace the `buildGetRequest` methods, etc. In general, it would be good to double check that the okhttp3 libs are leveraged idiomatically as much as possible.

I've also made some further clarification/refactor suggestions above. Some of the renaming would be good to get in this PR but larger suggestions could be handled separately.

lognaturel reviewed (changes requested):
Changes described above.

jd-alexander commented:
> From some quick reading, I think using Interceptors can make sharing e.g. headers between requests easier and that could probably replace the buildGetRequest methods, etc. In general, it would be good to double check that the okhttp3 libs are leveraged idiomatically as much as possible.

Spot on with this @lognaturel 

@xsteelej Normally, what I do is within the function that builds the `okHttpClient` I add all interceptors that represent common functionality that would live across multiple requests such as authentication and in this case `OpenRosa` headers. You could create something like `OpenRosaRequestInterceptor` that adds all the headers. 

As you go along, you might want to run several tests to ensure everything is working as expected and it might be difficult to be manually printing things to the console. There's this powerful interceptor that's available that generates a html page of a request and it's response so it can be shared and analyzed. It also prettifies the request/response before it goes to logcat. You can check it out here : https://github.com/simonpercic/OkLog

I also did a run through of the tests and I think since we are now using `okHttp` we can get a bit more extensive with our efforts so I am going to share some ideas on that shortly.

jd-alexander commented on collect_app/src/main/java/org/odk/collect/android/http/OkHttpConnection.java:
```diff
@@ -0,0 +1,352 @@
+    }
+
+    private OkHttpClient createOkHttpClient(@Nullable HttpCredentialsInterface credentials) {
+        OkHttpClient.Builder builder = new OkHttpClient.Builder();
```
@xsteelej it would be good if the `okHttpClient` was provided via Dagger so it would be easier to share it across multiple connections and also handle it's scope if the need arises.

xsteelej replied:
Hi @jd-alexander, I'm not sure if there would be an advantage in Dependency Injecting OkHttpClient - The intention of the Refactor was to isolate the underlying Http component in one place. Using Dagger would mean we'd have to start importing it else where etc.

lognaturel replied:
Agreed that the approach here is ok for now. I also propose some simplifications to how it's set in https://github.com/opendatakit/collect/compare/master...lognaturel:collect-620.

jd-alexander commented:
I am reading the OpenRosa auth spec [here](https://docs.opendatakit.org/openrosa-authentication/) and it states :
> Basic Authentication MUST NOT be performed over a non-secure (HTTP) connection.
Once the client is aware that basic authentication is required, it SHOULD proactively supply the basic authentication credentials on every secure request to the server, rather than wait for the server to reject the request with a 401 response.

What implications does this have for this refactor?

- Does it mean that we should check for `SSL` before trying to perform auth? 
- If it's not enabled how do we communicate this to the user? 
- How will the client be aware that basic auth is required? Would it be a case where once the username and password is saved then it just adds it to the request?

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/http/OkHttpConnection.java:
```diff
@@ -0,0 +1,352 @@
+                    .build();
+
+            builder.authenticator(new CachingAuthenticatorDecorator(authenticator, authCache))
+                    .addInterceptor(new AuthenticationCacheInterceptor(authCache));
```
I believe this takes care of pre-emptive auth and also handles the Ona case. What happens when there are two authenticators specified? Does it try the first and then the second if it's not supported? Does `DispatchingAuthenticator` take care of using the auth type specified in the auth challenge? It would be good to have a comment on the method with some of those details.

As @jd-alexander points out, there should also be a filter to only allow basic auth with SSL (the prior code filters for ports 443, 8443 which is an ok simplification). Throwing an exception which will eventually be shown to the user would be fine in the case of basic auth on HTTP.

lognaturel replied:
Had a quick private chat with @jd-alexander about testing and we thought the logic here would be worth adding tests for which I don't think exist yet. That is:
- If a server responds to a request with a 401 and requests digest auth, digest auth is used and the provided realm and nonce are used for further requests
- If an https server responds to a request with a 401 and requests basic auth, basic auth is used for further requests
- If an http server responds to a request with a 401 and requests basic auth without an https redirect, the client should not send auth headers
- The first `HEAD` request should not send auth headers. Once the client receives a 401 and finds out what auth is required by the server, then auth headers should be sent

I suspect that the interceptor as written will always send basic auth headers on the first `HEAD` request which shouldn't be the case. A test would help verify that either way.

xsteelej replied:
Hi @lognaturel,

As discussed, rather than writing tests I'm using proxy to intercept what happens so that I can answer the questions above.

_"If a server responds to a request with a 401 and requests digest auth, digest auth is used and the provided realm and nonce are used for further requests"_

**answer - Confirmed that the provided realm and nonce are used for further requests**

_"If an http server responds to a request with a 401 and requests basic auth without an https redirect, the client should not send auth headers"_

**answer - Added code that ensures that basic auth is not added if https is used.**

_"If an http server responds to a request with a 401 and requests basic auth without an https redirect, the client should not send auth headers"_

**answer - Was unable to test this, might need some help with configuring this scenario.**

_"The first HEAD request should not send auth headers. Once the client receives a 401 and finds out what auth is required by the server, then auth headers should be sent"_

**answer - Confirmed that the first head request does not send auth headers, but once a 401 is received then okhttp sends the relevant auth headers.**

xsteelej commented:
Hi @lognaturel & @jd-alexander, thanks for the comments. I'll start working through them and let you know if I need any clarification.

Thanks!

jrsteele09 committed:
- The version of okhttp3 used in the project is now a variable (Same version now used for implementation and test implementation)
- Added deflate back into the content encoding.
- OkHttpClient is now shared between requests

This is still work in progress, more review comments to address in a future update.

jrsteele09 committed:
Check Style and PMD error fixes.

jrsteele09 committed:
Fix for a "Dodgy Code Warning!" from spotbugs - Simplified the code.

jrsteele09 committed:
Removed the httpclient library + gradle dependencies include line.

jrsteele09 committed:
Removed the redundant check for headers.size being 0

jrsteele09 committed:
Renamed httpInterface.get to executeGetRequest and httpInterface.head to executeHeadRequest.

jrsteele09 committed:
Removed the unnecessary catch statements + reordered the statements.

jrsteele09 committed:
Updated tests

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge branch 'master' into issue_#620_PR2_break_dependency_on_httpclient

* master: (47 commits)
  Added scrollView in customDatePickerDialog (#2639)
  Remove file permission revoking for Android <=4.4 (#2638)
  Use toString() method instead of casting in order to avoid exceptions (#2636)
  Minor README updates (#2630)
  Fixed bug in filtering items that caused crashes (#2627)
  Fixed the disabled GeoWidget ImageButtons in Api21+. (#2276)
  Move uploadingComplete method up
  Pull out method that sends autosend results notification
  Clarify behavior of method that builds upload result message
  Extract method to get result messages
  Break up method to autosend instances
  Clarify when server polling happens
  Rename and comment to clarify semantics
  Rename and comment formShouldBeAutoSent method
  Nepali date widget (#2620)
  Remove undocumented ActivityLogger (#2589)
  Loading form didn't end when it contained an empty field-list group (#2612)
  Hide buttons if a question is readonly (#2563)
  Use FileProvider to open image/video in MediaLayout (#2605)
  Use FileProvider to access image files (qr code) (#2602)
  ...

# Conflicts:
#	collect_app/build.gradle

jrsteele09 committed:
As per review comment, removed try-catch block as it didn't really add anything worthwhile.

jrsteele09 committed:
Separated the functionality in ResponseMessageParser, the XML parsing remains in ResponseMessageParser and HttpPostResult contains the response and response code.

Added an executePostRequest to OpenRosaHttpInterface, implemented it in OkHttpConnection and refactored uploadSubmissionFile accordingly.

jrsteele09 committed:
Added executePostRequest() to MockHttpClientConnection

jrsteele09 committed:
PMD Error resolution

jrsteele09 committed:
- Renamed method getRequest, headRequest, postRequest to prepend them with "build"
- Ensured that "basic" authentication is not added if using https.

jrsteele09 committed:
Fixed PMD errors

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge branch 'master' into issue_#620_PR2_break_dependency_on_httpclient and fixed resulting build issues

* master: (74 commits)
  Look in google repo before jcenter
  Look in google repo before jcenter
  Use singleton database helper to avoid leaks
  Remove current() warning (#2668)
  Always allowClick across different activities (#2657)
  Fixed a bug in SelectmultipleWidget with autocomplete appearance (#2682)
  Support notification channels (#2676)
  Update ISSUE_TEMPLATE.md (#2680)
  Use bikram-sambat appearance instead of nepali (#2678)
  Close cursor after getting all forms from it
  Remove check for ongoing auto-send from GSheets Activity
  Describe the process for new features
  Set OSM map allowable min zoom to 2, max zoom to 22. (#2664)
  Update error message for missing submission URL in Google Sheets uploader (#2648)
  Improve comments for UploadException
  Move Outcome to InstanceUploaderTask and add comments
  Use try-with-resources
  Make location validating method static for easier testing
  Upgrade robolectric to 4.0-beta-1, add Guava
  Narrow down the HEAD request exception catching
  ...

# Conflicts:
#	build.gradle
#	collect_app/build.gradle
#	collect_app/src/androidTest/java/org/odk/collect/android/tasks/InstanceServerUploaderTaskTest.java
#	collect_app/src/main/java/org/odk/collect/android/tasks/InstanceServerUploader.java

jrsteele09 committed:
Fixed PMD errors

lognaturel commented:
@xsteelej Could you please share what the status of this PR is? Do you have a sense of when you might find another review useful?

xsteelej commented:
Hi @lognaturel

It would be useful for somebody to take a second look...

I'm currently working on the tests that yourself and @jd-alexander recommended but I have done most of the changes requested (I did have to sort some merge conflicts, so hopefully nothing has been undone 🤔). There are a few outstanding queries that I raised which may need looking at.

Also, I did try and squash the commit history, as recommended, but I ran into issues - I can try again once I've completed the tests.

Thanks

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/http/OkHttpConnection.java:
```diff
@@ -0,0 +1,357 @@
+    private static final String DATE_HEADER = "Date";
+    private static final String HTTP_CONTENT_TYPE_TEXT_XML = "text/xml";
+
+    private static OkHttpClient httpClient;
```
Is this only static because I manually built a `OkHttpConnection` in `AutoSendWorker`? Other than that, it should be an app-level singleton, right?

I did that because the whole Dagger setup is not quite right and I didn't want to keep going in the current direction (in particular, using `AppComponent` for everything). I don't exactly know what it should look like yet.

xsteelej replied:
I made it static to follow the recommended pattern of reusing the OkHttpClient object. If the httpClient exists then I just create a newBuilder from the httpClient to allow builder options to be modified. Otherwise, create a builder to create a new httpClient instance.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/http/OkHttpConnection.java:
```diff
@@ -0,0 +1,357 @@
+        return postResult;
+    }
+
+    private OkHttpClient createOkHttpClient(@Nullable HttpCredentialsInterface credentials, URI uri) {
```
This is an edge case but if the same credentials are used for an HTTP and HTTPS URI, the credentials may not be set correctly.

lognaturel replied:
A comment and rename would help explain what's going on here. See https://github.com/opendatakit/collect/commit/1b84188564d723b7b10b4afcf175aa43ee786a26. But as I wrote that comment, I got confused about why this method both mutates the client and returns it. Also, why build a new client here rather than in the constructor?

I tried what seems to me like a simpler approach in https://github.com/opendatakit/collect/commit/e0f37708aab211bf2638d9a7c013fb6aa6084a05, https://github.com/opendatakit/collect/commit/d40c47db507a3076ddd4564d20481b1ce6c82326 and https://github.com/opendatakit/collect/commit/56967692c19dc2ac13669e116cde650262cc4242 and seem to get the same behavior. Any reason you see not to do that? Did I correctly interpret the intent of the `httpCredentials` field?

xsteelej replied:
I agree, I'll refactor to make it more understandable and also take into account your code.

xsteelej replied:
Just as a follow up to my previous comment, createOkHttpClient used to do exactly that, but as previously pointed out, it's recommended to re-use the OkHttpClient. So this method has changed and now its name is confusing.

xsteelej replied:
Refactored [8a557ec](https://github.com/opendatakit/collect/pull/2595/commits/8a557ecfc09f0482da95aa6d26936f34bc78fb22)

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/http/OkHttpConnection.java:
```diff
@@ -0,0 +1,357 @@
+        addCredentials(builder, credentials, uri.getScheme().equalsIgnoreCase("https"));
+        httpCredentials = credentials;
+
+        httpClient = builder
```
Why does this need to be done again for a previously-configured builder? Can it just exist in the null `httpClient` case?

lognaturel replied:
Changed in https://github.com/opendatakit/collect/commit/e0f37708aab211bf2638d9a7c013fb6aa6084a05 and seems like it is not necessary.

xsteelej replied:
Due to some refactoring, and inspired by your code, I think this has been overcome by [8a557ec](https://github.com/opendatakit/collect/pull/2595/commits/8a557ecfc09f0482da95aa6d26936f34bc78fb22)

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/http/OkHttpConnection.java:
```diff
@@ -0,0 +1,357 @@
+            builder = new OkHttpClient.Builder();
+        }
+
+        addCredentials(builder, credentials, uri.getScheme().equalsIgnoreCase("https"));
```
This method does nothing in the case where the current request's credentials are null. Is that intentional? Should the credentials be cleared?

lognaturel replied:
I guess this probably doesn't matter because the server wouldn't ask for credentials if they're not needed?

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/http/OkHttpConnection.java:
```diff
@@ -0,0 +1,357 @@
+            return true;
+        } else if (httpCredentials == null || credentials == null) {
+            return false;
+        } else if (httpCredentials.isEqualTo(credentials)) {
```
Is there any particular reason not to use the standard `equals` method that all Java objects get? This seems pretty dangerous because users would expect the normal method to exist and it does but just returns the results of `==` AND equality methods typically check for null.

https://github.com/opendatakit/collect/commit/177306a5ec03d0ad396ea0d272de3051034cba6d is a commit that uses the standard equals.

xsteelej replied:
[8a557ec](https://github.com/opendatakit/collect/pull/2595/commits/8a557ecfc09f0482da95aa6d26936f34bc78fb22) uses standard `equals()` method.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/http/OkHttpConnection.java:
```diff
@@ -0,0 +1,357 @@
+
+    private void addCredentials(OkHttpClient.Builder builder, @Nullable HttpCredentialsInterface credentials, Boolean sslEnabled) {
+        if (credentials != null) {
+            final Map<String, CachingAuthenticator> authCache = new ConcurrentHashMap<>();
```
Could you please describe your understanding of the caching that occurs here? Since this cache is rebuilt every time `addCredentials` is run, I would expect it to only cache headers between requests with the same credentials. But I notice that in fact there's some unexpected behavior -- if I put in correct credentials in the server settings, do a request and then change the credentials to be wrong, the request still goes through. In fact, the latest and incorrect credentials should be used. Is this the level that caching occurs at? Somewhere else?

xsteelej replied:
I've investigated and it looks like the credentials will be cached unless you create a new http client. I've modified the code to recreate the http client if the credentials change or the scheme changes.

lognaturel commented:
Thanks for the update and your work on this, @xsteelej! I have asked some questions inline. Did I address all the queries you mentioned above?

So that we don't forget, I came across two defects:
- the credentials or perhaps headers are cached in an unexpected way. If I get a form listing with working credentials and then change the credentials in settings to bad credentials, requests still auth when they should fail. My question at https://github.com/opendatakit/collect/pull/2595#discussion_r230256561 might help clarify what's going on.
- getting a form listing from Central: `11-02 04:50:25.506 3712-3859/org.odk.collect.android E/CollectServerClient: Parsing failed with Unexpected token (position:TEXT �������������͘[o�V...@1:115 in java.io.InputStreamReader@264702b9)  while accessing https://odk.antinod.es/v1/key/<token>/formList`

steelej-arm committed:
Renamed isEqualTo to the standard Java equal

steelej-arm committed:
* Refactored of OkHttpClientConnection to cleanup the code that creates an OkHttpClient.
* Added standard equals method to HttpCredentials which replaces the previous method isEqualTo.

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge branch 'master' into issue_#620_PR2_break_dependency_on_httpclient

* master:
  Bump Dexter dependency to 5.0.0 (#2732)
  Handle case when the progress dialog needs to be dismissed after onSaveInstanceState (#2729)
  Use highContrastHighlight in grid widget (#2727)
  Improve naming of filtered form list including only newest (#2707)
  Fixed form styling in choices (#2723)
  Use production build with secret config to fix analytics mixing (#2708)
  Revert change from 77802cb4ca65864263a2fdfeb56288c9e571cd0d (#2724)
  Make rating widget stars orange (#2713)
  Fixed the trash button logic for map points, Fix #2633 (#2659)
  Fix error occurred when user adds file from SD card (#2654)
  Add rating widget (#2585)
  Provide guidance for WIP pull requests and add table of contents (#2702)
  Prevent hamcrest and rarepebble proguard warnings
  Upgrade to JR production build in preparation for release (#2699)
  Update October release translations
  Fixed temporary failed status when sending multiple submissions via SMS (#2692)
  Use query to get latest versions of forms (#2691)

zestyping commented:
Hi @xsteelej, @lognaturel is on leave now; can you leave a note here when you believe you have addressed all of her outstanding questions and comments, and then I'll take a final look?

xsteelej commented:
Hi @zestyping, thanks for your note and nice to meet you (virtually 😄)..

I'm just looking at the final bug which she spotted, which will also answer the question about my understanding of the caching as it's related.

Thanks

steelej-arm committed:
Modified the HttpCredentials equals() methods to call username and password string equals methods

Fixed bug with credentials being cached.

steelej-arm committed:
Fixed checkCode errors.

xsteelej commented:
Hi @zestyping, I've completed the code updates and I feel it is ready for a review.

Just a couple of points:

- I'm not sure what the bug is that @lognaturel ran into when it was reported that _"getting a form listing from Central: 11-02 04:50:25.506 3712-3859/org.odk.collect.android E/CollectServerClient: Parsing failed with Unexpected token.."_ - I was unable to reproduce.
- I haven't yet managed to test: _"If an http server responds to a request with a 401 and requests basic auth without an https redirect, the client should not send auth headers"_ - I could probably do with some pointers on how to setup https on my aggregate vm (I tried following an example about how to configure Tomcat for SSL but it didn't work).

xsteelej commented:
Hi @jd-alexander, do you have any time to take another look?

shobhitagarwal1612 commented:
@xsteelej I tried downloading blank forms & sending finalized forms using the default aggregate server and that worked perfectly.

xsteelej commented:
Thanks for testing @shobhitagarwal1612

grzesiek2010 commented:
@zestyping do you have some time for reviewing?

grzesiek2010 commented:
> I tried downloading blank forms & sending finalized forms using the default aggregate server and that worked perfectly.

@shobhitagarwal1612 
Did you also have a chance to review the code or it was just manual testing?

shobhitagarwal1612 commented:
> Did you also have a chance to review the code or it was just manual testing?

Yes, I did and it looks pretty clean to me.

grzesiek2010 commented:
@shobhitagarwal1612 

Thanks!
I'm going to review this pr as well and hopefully, @zestyping and @ggalmazor will find some time too.

yanokwa commented:
Hi folks! No need to review this one because I've asked @seadowg to review it. He is very familiar with OkHttp and once he looks at it, we can make a decision.

seadowg reviewed (changes requested):
Hey @xsteelej! I managed to have a look a this this evening. My interpretation of the aim of this PR is to switch the implementation of `OpenRosaHttpInterface` from `HttpClientConnection` to a new `OkHttpConnection` (that uses OkHttp). Given that goal I couldn't spot any problems in the code and the addition of OkHttp looks good. There are two thing I'd like to suggest though:

1. It is quite hard to to be confident that no minor changes in behavior have been introduced as there were no existing tests at an HTTP integration level for `HttpClientConnection` as far as I can see. I'll look into adding these so we can verify there are no minor differences between the two implementations of the interface. Given this I think what makes most sense is to change this PR so it doesn't remove the ol' `HttpClientConnection` (or its uses) but simply adds the new `OkHttpConnection` (which will be unused for the moment).

2. The Git commit history contains quite a lot of merge commits. Would it be possible to try and clean these up? I'd imagine it's possible to squash the commits down so the history is more readable and then force push the branch. As always SO to the rescue: https://stackoverflow.com/questions/5189560/squash-my-last-x-commits-together-using-git

xsteelej commented:
Hi @seadowg, thanks for the review!

1. What are your thoughts @yanokwa on the suggestion to merge this OkHttpClient and leave it unused for now? Another alternative could be that we merge it early in a release cycle and have it tested early.

2. I'll try and clean up the merge commits.. I did try, due to some previous comments, a while back and it didn't work out too well. However, I'll give it another go.

Thanks

yanokwa commented:
@xsteelej I think merging unused for now is the right approach. It just gives us more flexibility as far as when exactly we turn it on. As to cleaning up the commits, you can make a branch off this current one to play around with with interactive rebasing.

https://robots.thoughtbot.com/git-interactive-rebase-squash-amend-rewriting-history has a pretty good guide. https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History is also pretty good.

yanokwa commented:
@xsteelej I'd like to get the requested changes merged a few weeks before our Feb 20th code freeze. Is that possible?

xsteelej commented:
Hi @yanokwa, it should be yes - I've managed to squash most of the merge upstream/master commits on a branch, just looking at the latest conflicts and then i'll put back the httpConnection.

Thanks
John

xsteelej commented:
To remove all the "Merge remote-tracking branch 'upstream/master'" commits I have created a new pull request: https://github.com/opendatakit/collect/pull/2849 which replaces this one.

The new PR adds okhttp3 but it is not currently used.

grzesiek2010 commented:
closed in favor of https://github.com/opendatakit/collect/pull/2849
